driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/ClientSessionPublisherImpl.java

```diff
@@ -74,6 +74,13 @@ public boolean notifyMessageSent() {
         }
     }
 
+    @Override
+    public void notifyNonCommitOperationInitiated() {
+        if (transactionState == TransactionState.COMMITTED) {
```

Comment:
I understand that we cannot rely solely on `transactionState` and need additional knowledge that the operation is not a `CommitTransactionOperation` because commit may be called multiple times. However, I do not understand why we check for `transactionState == TransactionState.COMMITTED` instead of checking for `transactionState != TransactionState.IN` to determine whether an operation is executed in a transaction. @jyemin, could you please clarify this?

Reply:
It's because if state is IN or ABORTED, no server will be pinned, so it would be a no-op.  I thought this way would make the intent clearer.  I see three choices

1. Leave it as is
2. Add an assertion that nothing is pinned if state is ABORTED or NONE (there should already be tests that assert this)
3. Change conditional to `transactionState != IN`

What's the preference?

Reply:
Thanks @jyemin makes sense, I'm happy for it to be kept as is. I think part of the confusion is the name of the method, versus what it does. Could we make that clearer?

Reply:
I was trying to keep it in the spirit of the existing `notifyMessageSent`.  What do you suggest as an alternative name?

Reply:
`unPinServerIfNotCommitted` ?

Reply:
Happy for it to be left as well.

Reply:
> keep it in the spirit of the existing notifyMessageSent

It seems best to have a method `notifyOperationInitiated` with all the logic that decides which operations to react and how in the method. Other approaches with necessity move pieces of this logic out of the method to the surrounding code, and I do not see a good reason to break the encapsulation this way.

> Add an assertion that nothing is pinned if state is ABORTED or NONE

Yes, having
```java
if (transactionState != IN) {
  assertTrue(pinnedServerAddress == null || (transactionState != ABORTED && transactionState != NONE));//optional
  pinnedServerAddress = null;
}
```
makes the intent of the `if` statement much clearer. Relying here on the not obvious knowledge _"if state is IN or ABORTED, no server will be pinned"_ makes the code hard to understand.

I also noticed that there is a method `hasActiveTransaction`. Is it possible that if we use this method, we can stop checking whether or not an action is a commit action?

Reply:
Added the requested assertion logic.

The `hasActiveTransaction` method, while tempting, does not serve the required purpose, as the first time commitTransaction is executed the state is `IN`, but we don't want to unpin.

Given this, I'm not sure if you would like to see further changes, so please let me know.  Happy to go further if it makes sense.

Reply:
I have two proposals, which may require a discussion.

**1. Improve encapsulation**
Why do we move the logic of checking the operation type outside of `ClientSessionPublisherImpl` by introducing method `notifyNonCommitOperationInitiated` instead of having `notifyOperationInitiated` that would check the operation type itself? I propose having `notifyOperationInitiated` method instead.

**2. Use `!hasActiveTransaction()` instead of `transactionState != IN`**
> The `hasActiveTransaction` method, while tempting, does not serve the required purpose, as the first time commitTransaction is executed the state is `IN`, but we don't want to unpin.

With `if (!hasActiveTransaction())` we cannot unpin in the `IN` state because the method returns `true` for this state.

If we compare `if (transactionState != IN)` vs `if (!hasActiveTransaction())`, the comparison is equivalent to `transactionState != IN` vs `transactionState != IN && !(transactionState == COMMITTED && commitInProgress)`. The latter is a more strict condition, which does not match situations `transactionState == TransactionState.COMMITTED && commitInProgress`. `commitInProgress` is only true while a commit operation is being executed, which means that no other operation may be initiated concurrently with `commitInProgress` being true. Therefore, for the purpose of the `notifyNonCommitOperationInitiated` (or `notifyOperationInitiated`) method the checks `if (transactionState != IN)` and `if (!hasActiveTransaction())` are equivalent, with the latter being more expressive of its intent and using a check that already exists instead of introducing another very similar one.

Reply:
Agreed on all points.  I was thinking if it were to become operation-aware, I would have to to add two overloads, one that took a ReadOperation and one that took a WriteOperation, and even worse those interfaces are not in the public API.  The thing I didn't see, which is obvious now, is that it's ok for this new method to take an Object.  

Just to be safe, I added an assertion that the Object is either an instance of ReadOperation or WriteOperation.